File: sdks/node/src/api/endpoint.ts
import {
  type EndpointHeadersIn,
  EndpointHeadersInSerializer,
} from "../models/endpointHeadersIn";
import {
  type EndpointHeadersOut,
  EndpointHeadersOutSerializer,
} from "../models/endpointHeadersOut";
import {
  type EndpointHeadersPatchIn,
  EndpointHeadersPatchInSerializer,
} from "../models/endpointHeadersPatchIn";
import { type EndpointIn, EndpointInSerializer } from "../models/endpointIn";
import { type EndpointOut, EndpointOutSerializer } from "../models/endpointOut";
import { type EndpointPatch, EndpointPatchSerializer } from "../models/endpointPatch";
import {
  type EndpointSecretOut,
  EndpointSecretOutSerializer,
} from "../models/endpointSecretOut";
import {
  type EndpointSecretRotateIn,
  EndpointSecretRotateInSerializer,
} from "../models/endpointSecretRotateIn";
import {
  type EndpointUpdate,
  EndpointUpdateSerializer,
} from "../models/endpointUpdate";
import {
  type ListResponseEndpointOut,
  ListResponseEndpointOutSerializer,
} from "../models/listResponseEndpointOut";
import { HttpMethod, HookSniffRequest, type HookSniffRequestContext } from "../request";

export interface EndpointListOptions {
  limit?: number;
  iterator?: string | null;
  order?: string;
}

export interface EndpointCreateOptions {
  idempotencyKey?: string;
}

export interface EndpointRotateSecretOptions {
  idempotencyKey?: string;
}

export class Endpoint {
  public constructor(private readonly requestCtx: HookSniffRequestContext) {}

  /** List endpoints. */
  public list(options?: EndpointListOptions): Promise<ListResponseEndpointOut> {
    const request = new HookSniffRequest(HttpMethod.GET, "/api/v1/endpoint");

    request.setQueryParams({
      limit: options?.limit,
      iterator: options?.iterator,
      order: options?.order,
    });

    return request.send(
      this.requestCtx,
      ListResponseEndpointOutSerializer._fromJsonObject
    );
  }

  /** Create a new endpoint. */
  public create(
    endpointIn: EndpointIn,
    options?: EndpointCreateOptions
  ): Promise<EndpointOut> {
    const request = new HookSniffRequest(HttpMethod.POST, "/api/v1/endpoint");

    request.setHeaderParam("idempotency-key", options?.idempotencyKey);
    request.setBody(EndpointInSerializer._toJsonObject(endpointIn));

    return request.send(this.requestCtx, EndpointOutSerializer._fromJsonObject);
  }

  /** Get an endpoint. */
  public get(endpointId: string): Promise<EndpointOut> {
    const request = new HookSniffRequest(
      HttpMethod.GET,
      "/api/v1/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.send(this.requestCtx, EndpointOutSerializer._fromJsonObject);
  }

  /** Update an endpoint. */
  public update(
    endpointId: string,
    endpointUpdate: EndpointUpdate
  ): Promise<EndpointOut> {
    const request = new HookSniffRequest(
      HttpMethod.PUT,
      "/api/v1/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setBody(EndpointUpdateSerializer._toJsonObject(endpointUpdate));

    return request.send(this.requestCtx, EndpointOutSerializer._fromJsonObject);
  }

  /** Delete an endpoint. */
  public delete(endpointId: string): Promise<void> {
    const request = new HookSniffRequest(
      HttpMethod.DELETE,
      "/api/v1/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Partially update an endpoint. */
  public patch(
    endpointId: string,
    endpointPatch: EndpointPatch
  ): Promise<EndpointOut> {
    const request = new HookSniffRequest(
      HttpMethod.PATCH,
      "/api/v1/endpoint/{endpoint_id}"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setBody(EndpointPatchSerializer._toJsonObject(endpointPatch));

    return request.send(this.requestCtx, EndpointOutSerializer._fromJsonObject);
  }

  /** Get the additional headers to be sent with the webhook. */
  public getHeaders(endpointId: string): Promise<EndpointHeadersOut> {
    const request = new HookSniffRequest(
      HttpMethod.GET,
      "/api/v1/endpoint/{endpoint_id}/headers"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.send(this.requestCtx, EndpointHeadersOutSerializer._fromJsonObject);
  }

  /** Set the additional headers to be sent with the webhook. */
  public updateHeaders(
    endpointId: string,
    endpointHeadersIn: EndpointHeadersIn
  ): Promise<void> {
    const request = new HookSniffRequest(
      HttpMethod.PUT,
      "/api/v1/endpoint/{endpoint_id}/headers"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setBody(EndpointHeadersInSerializer._toJsonObject(endpointHeadersIn));

    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Partially set the additional headers to be sent with the webhook. */
  public patchHeaders(
    endpointId: string,
    endpointHeadersPatchIn: EndpointHeadersPatchIn
  ): Promise<void> {
    const request = new HookSniffRequest(
      HttpMethod.PATCH,
      "/api/v1/endpoint/{endpoint_id}/headers"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setBody(EndpointHeadersPatchInSerializer._toJsonObject(endpointHeadersPatchIn));

    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Get the endpoint's signing secret. */
  public getSecret(endpointId: string): Promise<EndpointSecretOut> {
    const request = new HookSniffRequest(
      HttpMethod.GET,
      "/api/v1/endpoint/{endpoint_id}/secret"
    );

    request.setPathParam("endpoint_id", endpointId);

    return request.send(this.requestCtx, EndpointSecretOutSerializer._fromJsonObject);
  }

  /** Rotate the endpoint's signing secret. */
  public rotateSecret(
    endpointId: string,
    endpointSecretRotateIn: EndpointSecretRotateIn,
    options?: EndpointRotateSecretOptions
  ): Promise<void> {
    const request = new HookSniffRequest(
      HttpMethod.POST,
      "/api/v1/endpoint/{endpoint_id}/secret/rotate"
    );

    request.setPathParam("endpoint_id", endpointId);
    request.setHeaderParam("idempotency-key", options?.idempotencyKey);
    request.setBody(EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));

    return request.sendNoResponseBody(this.requestCtx);
  }
}

File: sdks/rust/src/api/endpoint.rs
// this file is @generated
use crate::{error::Result, models::*, Configuration};

#[derive(Default)]
pub struct EndpointListOptions {
    pub limit: Option<i32>,
    pub iterator: Option<String>,
    pub order: Option<Ordering>,
}

#[derive(Default)]
pub struct EndpointCreateOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointRotateSecretOptions {
    pub idempotency_key: Option<String>,
}

#[derive(Default)]
pub struct EndpointSendExampleOptions {
    pub idempotency_key: Option<String>,
}

pub struct Endpoint<'a> {
    cfg: &'a Configuration,
}

impl<'a> Endpoint<'a> {
    pub(super) fn new(cfg: &'a Configuration) -> Self {
        Self { cfg }
    }

    pub async fn list(
        &self,
        app_id: String,
        options: Option<EndpointListOptions>,
    ) -> Result<ListResponseEndpointOut> {
        let EndpointListOptions { limit, iterator, order } = options.unwrap_or_default();
        crate::request::Request::new(http1::Method::GET, "/api/v1/app/{app_id}/endpoint")
            .with_path_param("app_id", app_id)
            .with_optional_query_param("limit", limit)
            .with_optional_query_param("iterator", iterator)
            .with_optional_query_param("order", order)
            .execute(self.cfg)
            .await
    }

    pub async fn create(
        &self,
        app_id: String,
        endpoint_in: EndpointIn,
        options: Option<EndpointCreateOptions>,
    ) -> Result<EndpointOut> {
        let EndpointCreateOptions { idempotency_key } = options.unwrap_or_default();
        crate::request::Request::new(http1::Method::POST, "/api/v1/app/{app_id}/endpoint")
            .with_path_param("app_id", app_id)
            .with_optional_header_param("idempotency-key", idempotency_key)
            .with_body_param(endpoint_in)
            .execute(self.cfg)
            .await
    }

    pub async fn get(&self, app_id: String, endpoint_id: String) -> Result<EndpointOut> {
        crate::request::Request::new(http1::Method::GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .execute(self.cfg)
            .await
    }

    pub async fn update(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_update: EndpointUpdate,
    ) -> Result<EndpointOut> {
        crate::request::Request::new(http1::Method::PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .with_body_param(endpoint_update)
            .execute(self.cfg)
            .await
    }

    pub async fn delete(&self, app_id: String, endpoint_id: String) -> Result<()> {
        crate::request::Request::new(http1::Method::DELETE, "/api/v1/app/{app_id}/endpoint/{endpoint_id}")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .returns_nothing()
            .execute(self.cfg)
            .await
    }

    pub async fn patch(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_patch: EndpointPatch,
    ) -> Result<EndpointOut> {
        crate::request::Request::new(http1::Method::PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .with_body_param(endpoint_patch)
            .execute(self.cfg)
            .await
    }

    pub async fn get_headers(&self, app_id: String, endpoint_id: String) -> Result<EndpointHeadersOut> {
        crate::request::Request::new(http1::Method::GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .execute(self.cfg)
            .await
    }

    pub async fn update_headers(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_headers_in: EndpointHeadersIn,
    ) -> Result<()> {
        crate::request::Request::new(http1::Method::PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .with_body_param(endpoint_headers_in)
            .returns_nothing()
            .execute(self.cfg)
            .await
    }

    pub async fn patch_headers(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_headers_patch_in: EndpointHeadersPatchIn,
    ) -> Result<()> {
        crate::request::Request::new(http1::Method::PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .with_body_param(endpoint_headers_patch_in)
            .returns_nothing()
            .execute(self.cfg)
            .await
    }

    pub async fn get_secret(&self, app_id: String, endpoint_id: String) -> Result<EndpointSecretOut> {
        crate::request::Request::new(http1::Method::GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .execute(self.cfg)
            .await
    }

    pub async fn rotate_secret(
        &self,
        app_id: String,
        endpoint_id: String,
        endpoint_secret_rotate_in: EndpointSecretRotateIn,
        options: Option<EndpointRotateSecretOptions>,
    ) -> Result<()> {
        let EndpointRotateSecretOptions { idempotency_key } = options.unwrap_or_default();
        crate::request::Request::new(http1::Method::POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret/rotate")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .with_optional_header_param("idempotency-key", idempotency_key)
            .with_body_param(endpoint_secret_rotate_in)
            .returns_nothing()
            .execute(self.cfg)
            .await
    }

    pub async fn send_example(
        &self,
        app_id: String,
        endpoint_id: String,
        event_example_in: EventExampleIn,
        options: Option<EndpointSendExampleOptions>,
    ) -> Result<MessageOut> {
        let EndpointSendExampleOptions { idempotency_key } = options.unwrap_or_default();
        crate::request::Request::new(http1::Method::POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/send-example")
            .with_path_param("app_id", app_id)
            .with_path_param("endpoint_id", endpoint_id)
            .with_optional_header_param("idempotency-key", idempotency_key)
            .with_body_param(event_example_in)
            .execute(self.cfg)
            .await
    }
}
